### ISSUE TITLE
unreachable code error: parameters not get rewritten in extended mode

### ISSUE THREAD

**Issue by rotten** [risingwavelabs/risingwave#15094]
### Describe the bug

I ran into this on the frontend, and since it said I should file a report, I did.   I'm not sure exactly what query I was running to trip it.

### Error message/log

```text
risingwave-frontend-1  | thread 'rw-main' panicked at src/frontend/src/optimizer/plan_expr_rewriter/const_eval_rewriter.rs:46:43:
risingwave-frontend-1  | internal error: entered unreachable code: Parameter should not appear here. It will be replaced by a literal before this step.
risingwave-frontend-1  | 2024-02-14T19:46:14.991239159Z ERROR handle_query{mode="extended query execute" session_id=4 sql=SELECT EXISTS (SELECT true FROM rw_schemas AS s JOIN rw_tables AS t ON s.id = t.schema_id WHERE s.name = $1 AND t.name = $2), params = Some([None, None])}: pgwire::pg_protocol: error when process message error=Panicked when handling the request: internal error: entered unreachable code: Parameter should not appear here. It will be replaced by a literal before this step.
risingwave-frontend-1  | This is a bug. We would appreciate a bug report at:
risingwave-frontend-1  |   https://github.com/risingwavelabs/risingwave/issues/new?labels=type%2Fbug&template=bug_report.yml
```

### To Reproduce

I'm not sure.  I was running a script.  Every time I kick it off, it throws this.   If I can chase down exactly what the script is doing I'll update this report.

### Expected behavior

the query should succeed.  When I run the one in the error log, and swap out the params with nulls, it works:
```
SELECT EXISTS (SELECT true FROM rw_schemas AS s JOIN rw_tables AS t ON s.id = t.schema_id WHERE s.name = null AND t.name = null)
```

### How did you deploy RisingWave?

docker-compose

### The version of RisingWave

```
localhost root@dev=> select version();
                                      version
------------------------------------------------------------------------------------
 PostgreSQL 9.5.0-RisingWave-1.7.0-alpha (85f002347c65b3a955ff110146629f1d06890b48)
(1 row)

```

**Comment by rotten** [comment]
Double checking, the query my script was running, was in fact, the one in the error log.  However it was passing parameters, for whatever reason, the error log replaced those values with `None`.

**Comment by BugenZhao** [comment]
Hi and thanks for your feedback.

It appears that there's something wrong with the execution flow of a parameterized query in the extended query mode.

Would you mind providing a minimal portion of your script so that we can reproduce the issue?

cc @ZENOTME

**Comment by rotten** [comment]
I'm trying to make a risingwave Engine for sqitch (https://sqitch.org).  This query is my refactoring of the 'initialized' function:
```
sub initialized {
    my $self = shift;
    return $self->dbh->selectcol_arrayref(q{
               SELECT EXISTS(
                   SELECT
                        TRUE
                    FROM
                        rw_schemas s
                        join rw_tables t on s.id = t.schema_id
                     WHERE
                         s.name = ?
                         AND
                         t.name = ?
                )
        }, undef, $self->registry, 'changes')->[0];
}
```
It is perl.  I'm about to give up on that project because my perl skills are too lame to do it well.   This function is called when I run `sqitch log`.  It doesn't seem to matter if sqitch is initialized or not - it always throws this error.

FWIW, the original block of code looked like this: https://github.com/sqitchers/sqitch/blob/fa315390f2f60aeef6c891aeafae7ebbd3da461e/lib/App/Sqitch/Engine/pg.pm#L200

**Comment by ZENOTME** [comment]
Sorry, I think it caused by I forget to rewrite the subquery in parameter rewriter. Fix it in #15095
